docs/elementExtensions.md

```diff
@@ -0,0 +1,73 @@
+# Element extensions developer guide
+
+Extensions are a way to add custom logic and extra complexity to already existing elements.  Each element has the ability to load relevant extensions and as such, extensions should be custom tailored to one specific element.
+
+## Extension anatomy
+
+Extensions can be defined in a course only, and are located in `[course directory]/elementExtensions`.  This folder is organized such that elements are contained as subfolders, and extensions are then contained inside the element folders.  For example, if you had an extension `exampleExtension` that extended `exampleElement`, the folder structure would be: `[course directory]/elementExtensions/exampleElement/exampleExtension`.
+
+Each extension needs only an `info.json` containing metadata about the element, and can contain optional information like Python scripts, CSS styles, and clientside JavaScript files.
+
+The `info.json` file is structurally similar to the element info file and may contain the following fields:
+```json
+{
+    "controller": "Python script",
+    "dependencies": {
+        "coreStyles": ["Core PL styles"],
+        "coreScripts": ["Core PL JavaScripts"],
+        "nodeModulesStyles": ["Node modules styles"],
+        "nodeModulesScripts": ["Node modules JavaScripts"],
+        "clientFilesCourseStyles": ["Client files course styles"],
+        "clientFilesCourseScripts": ["Client files course JavaScripts"],
+        "extensionStyles": ["Styles in the extension directory"],
+        "extensionScripts": ["JavaScripts in the extension directory"]
+    }
+}
+```
+
+### Python Controller
+
+The main Python controller script has no general structure and is instead defined by the element that is being extended.  Any global functions and variables are available for use by the host element.
+
+A host element can call the `load_element_extension()` function to load one specific extension, or `load_all_extensions_for_element()` to load everything that is available.  These are defined in the freeform `prairielearn` module.
+
+Loading extension Python scripts returns a named tuple of all globally defined functions and variables.  Loading all extensions will return a dictionary mapping the extension name to its named tuple.  For example, if an extension were to define the following in their controller:
+
+```python
+def my_cool_function():
+    return "hello, world!"
+```
+
+The host element could then call this by running the following:
+
+```python
+import prairielearn as pl
+
+def render(element_html, data):
+    extension = pl.load_element_extensions(data, 'extension_name')
+    contents = extension.my_cool_function()
+    return contents
+```
+
+This small example above will render `"hello world!"` to the question page.  Note that no order is guaranteed when loading all extensions with `load_all_extensions_for_element()`.
```

Comment:
I'm not sure allowing for a random order is a good thing... Is the reason why this is the case because `prairielearn` is using a version of `Python` before 3.6 so `dict` structures are missing order? If so, could we consider adding `OrdDict` from `collections` ala `from collections import OrderedDict`?

Details: https://stackoverflow.com/a/39537308/1345455

Reply:
Sure, though I'm not sure the order really matters as long as it's consistent?  I can sort the extensions in alphabetical order so it's deterministic.